### ISSUE TITLE
Provide a way to propagate reload requests from targets to services

### ISSUE THREAD

**Issue by mbiebl** [systemd/systemd#710]
This is a followup from the mailing list discussion http://lists.freedesktop.org/archives/systemd-devel/2015-July/033633.html

2015-07-22 20:41 GMT+02:00 Lennart Poettering lennart@poettering.net:

> On Wed, 22.07.15 20:28, Michael Biebl (mbiebl@gmail.com) wrote:
> 
> > 2015-07-22 19:15 GMT+02:00 Lennart Poettering lennart@poettering.net:
> > 
> > > On Tue, 21.07.15 13:43, Marc Haber (mh+systemd-devel@zugschlus.de) wrote:
> > > 
> > > > Can I write my nifty.target as a service? I have seen in this case
> > > > nifty.service files with Exec=/bin/true to basically create a no-op
> > > > service, but that's ugly.
> > > 
> > > That's a hack. A target is for grouping things, they basically are
> > > services without processes attached to them. You really should use
> > > targets for this, and not misuse services.
> > 
> > Afaik, there is no way to have a .target reloaded and that reload
> > request propagated to all services in that target.
> > 
> > Say you have foo.target consisting of  bar.service and baz.service,
> > how can you achieve that
> > systemctl reload foo.target propagates that reload request to {bar,baz}.service?
> 
> You cannot do that right now, and I am not sure you should be able to,
> after all the operation does not really apply to target units.
> 
> I do see your usecase though...
> 
> Humm...
> 
> Maybe we can change the manager core to propagate Reload() calls
> for unit type that do not support it natively to other units listed in
> PropagateReloadsTo= and then become a NOP.
> 
> Or in other words: invoking reload on a target that knows no
> PropagateReloadsTo= should continue to return an error. But if such
> deps are defined, it should become a silent NOP and propagate the
> event.
> 
> Could you file an RFP issue on github asking for this? [or even better,
> send a PR with a patch...  ;-) ]

**Comment by df7cb** [comment]
We have the same problem for PostgreSQL in Debian. There can be an arbitrary number of PostgreSQL "clusters" (PG speak for database instance) running, managed by postgresql@.service. In the past, /etc/init.d/postgresql start/stop/restart/reload would act on all clusters. To get the same behavior with systemd, I started with making a postgresql.target, but quickly realized that targets do not support reload. Now we have a postgresql.service with Exec=/bin/true which works, but which keeps confusing users because using true there is just weird.

PropagateReloadsTo=<static something> won't work because the set of clusters is dynamic. Currently we have this in postgresql@.service:

```
PartOf=postgresql.service
ReloadPropagatedFrom=postgresql.service
```

IMHO the existing PartOf and ReloadPropagatedFrom settings would just work if targets would accept them. (I can see that you wouldn't want "systemctl reload default.target" to reload the whole system, but if something explicitely says "ReloadPropagatedFrom some.target", that could Just Work and wouldn't violate the POLA.)

Thanks for considering :)

**Comment by mbiebl** [comment]
Let me copy my reply from #999:

I think I'd like to see a boolean PropagateReload=yes instead or at least in addition to PropagateReloadTo=foo.service
This way services can declare themselves PartOf=foo.target, which can be useful for instanced services, where you don't know the name of the service beforehand.
would basically use the same definition here that we use for restart.

**Comment by poettering** [comment]
BTW, this will be fixed by #6428.

**Comment by poettering** [comment]
Closing as #6428 got merged.

**Comment by martinpitt** [comment]
Reopening, as #6428 caused a regression and was reverted in #6836 .

**Comment by johnlinp** [comment]
Any plan or milestone to fix this issue?

**Comment by boucman** [comment]
well, the half of the original PR that fixed this has been comitted and released in v235

what's left of the PR is the part that causes a .device reload on event. So I think this FR can actually be closed...

**Comment by johnlinp** [comment]
@martinpitt What do you say? Thanks.

**Comment by poettering** [comment]
Implemented in f54bcca5c16bea2ed6b694c7944cb0ae679fbfaf. Closing

**Comment by johnlinp** [comment]
@poettering Thank you for confirming this.
